File: android/src/main/java/com/bazumax/capacitor/firebase/storage/FirebaseStoragePlugin.kt
package com.bazumax.capacitor.firebase.storage

import android.app.Activity
import android.net.Uri
import androidx.activity.result.ActivityResult
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.ActivityCallback
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.util.*


@CapacitorPlugin(name = "FirebaseStorage")
public class FirebaseStoragePlugin : Plugin() {
    private val implementation: FirebaseStorage = FirebaseStorage()

    override fun load() {
        super.load()
        implementation.load()
    }
    fun sendState(state: String) {
        val obj = JSObject()
        obj.put("state", state)
        notifyListeners("onUploadStateChanged", obj)
    }

    fun selectImage(call: PluginCall) {
        val title: String = call.getString("title", "Выберите аватар")!!

        val intent = implementation.createSelectImageRequest(title)
        startActivityForResult(call, intent, "selectImageCallback")

    }

    suspend fun compressThenUpload(call: PluginCall, uri: Uri) {
        val compressedFileUri = comressImage(uri, activity)
        uploadRealFile(call, compressedFileUri)
    }

    fun uploadRealFile(call: PluginCall, uri: Uri) {
        val storagePath = call.getString("storagePath", "avatar")!!

        sendState("uploading")
        implementation.uploadFile(uri, storagePath) { result: Boolean, uriOrMessage: String ->
            if (result) {
                val obj = JSObject()
                obj.put("url", uriOrMessage)
                call.resolve(obj)
            } else {
                call.reject(uriOrMessage)
            }
        }
    }

    @ActivityCallback
    fun selectImageCallback(call: PluginCall, result: ActivityResult) {
        if (result.resultCode != Activity.RESULT_OK) {
            call.reject("Invalid result code: " + result.resultCode)
            return
        }

        var uri = result.data!!.data!!

        GlobalScope.launch(Dispatchers.Main) {
            sendState("compressing")
            if (call.getBoolean("compress", false) == true) {
                compressThenUpload(call, uri)
                return@launch
            }

            uploadRealFile(call, uri)
        }
    }

    @PluginMethod
    public fun uploadFile(call: PluginCall) {
        if (call.getBoolean("withImagePicker") == true) {
            selectImage(call)
            return
        }

        // TODO: base upload
    }

    @PluginMethod
    public fun getDownloadUrl(call: PluginCall) {
        val path = call.getString("path")
        if (path == null) {
            call.reject("path is null")
            return
        }
        implementation.getDownloadUrl(path) { result: Boolean, uriOrMessage: String ->
            if (result) {
                val obj = JSObject()
                obj.put("url", uriOrMessage)
                call.resolve(obj)
            } else {
                call.reject(uriOrMessage)
            }
        }
    }

    @PluginMethod
    public fun deleteFile(call: PluginCall) {
        val path = call.getString("path")
        if (path == null) {
            call.reject("path is null")
            return
        }

        implementation.deleteFile(path) { result: Boolean, message: String ->
            if (result) {
                val obj = JSObject()
                obj.put("message", message)
                call.resolve(obj)
            } else {
                call.reject(message)
            }
        }
    }

    @PluginMethod
    public fun watchFileCreation(call: PluginCall) {
        val path = call.getString("path")
        if (path == null) {
            call.reject("path is null")
            return
        }

        implementation.watchFileCreation(path) { result: Boolean, urlOrMessage: String ->
            if (result) {
                val obj = JSObject()
                obj.put("url", urlOrMessage)
                call.resolve(obj)
            } else {
                call.reject(urlOrMessage)
            }
        }
    }
}

File: ios/Plugin/FirebaseStoragePlugin.swift
import Foundation
import Capacitor
/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(FirebaseStoragePlugin)
public class FirebaseStoragePlugin: CAPPlugin {
    private let implementation = FirebaseStorageManager()

    func sendState(state: String) {
        self.notifyListeners("onUploadStateChanged", data: ["state": state])
    }

    @objc func uploadFile(_ call: CAPPluginCall) {
        let storagePath = call.getString("storagePath") ?? ""
        if (call.getBool("withImagePicker", false) == true) {
            implementation.selectImage(view: (bridge?.viewController)!) { (result, url) in
                if (!result) {
                    call.reject("User cancelled")
                    return
                }

                self.sendState(state: "uploading")
                self.implementation.uploadFile(path: storagePath, url: url!, contentType: "image/jpeg", self.resolveCallback(key: "url", call: call))
            }
        }
    }


    @objc func deleteFile(_ call: CAPPluginCall) {
        let path = call.getString("path") ?? ""
        implementation.deleteFile(path: path, resolveCallback(key: "message", call: call))
    }

    @objc func getDownloadUrl(_ call: CAPPluginCall) {
        let path = call.getString("path") ?? ""
        implementation.getDownloadUrl(path: path, resolveCallback(key: "url", call: call))
    }

    @objc func watchFileCreation(_ call: CAPPluginCall) {
        let path = call.getString("path") ?? ""
        implementation.watchFileCreation(path: path, resolveCallback(key: "url", call: call))
    }

    func resolveCallback(key: String, call: CAPPluginCall) -> (((Bool, String?) -> ())) {
        return { result, value in
            if (!result) {
                call.reject("Invalid code: 1")
                return
            }

            call.resolve([
                key: value!
            ])
        }
    }
}
